internal/component/prometheus/remote/queue/serialization/serializer.go

```diff
@@ -0,0 +1,223 @@
+		metaInbox:           actor.NewMailbox[*types.TimeSeriesBinary](),
+		cfgInbox:            actor.NewMailbox[types.SerializerConfig](),
+		flushTestTimer:      time.NewTicker(1 * time.Second),
+		msgpBuffer:          make([]byte, 0),
+		lastFlush:           time.Now(),
+		stats:               stats,
+	}
+
+	return s, nil
+}
+func (s *serializer) Start() {
+	// All the actors and mailboxes need to start.
+	s.queue.Start()
+	s.self = actor.Combine(actor.New(s), s.inbox, s.metaInbox, s.cfgInbox).Build()
+	s.self.Start()
+}
+
+func (s *serializer) Stop() {
+	s.queue.Stop()
+	s.self.Stop()
+}
+
+func (s *serializer) SendSeries(ctx context.Context, data *types.TimeSeriesBinary) error {
+	return s.inbox.Send(ctx, data)
+}
+
+func (s *serializer) SendMetadata(ctx context.Context, data *types.TimeSeriesBinary) error {
+	return s.metaInbox.Send(ctx, data)
+}
+
+func (s *serializer) UpdateConfig(ctx context.Context, cfg types.SerializerConfig) error {
+	return s.cfgInbox.Send(ctx, cfg)
+}
+
+func (s *serializer) DoWork(ctx actor.Context) actor.WorkerStatus {
+	// Check for config which should have priority. Selector is random but since incoming
+	// series will always have a queue by explicitly checking the config here we always give it a chance.
+	// By pulling the config from the mailbox we ensure it does NOT need a mutex around access.
+	select {
+	case <-ctx.Done():
+		return actor.WorkerEnd
+	case cfg, ok := <-s.cfgInbox.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		s.maxItemsBeforeFlush = int(cfg.MaxSignalsInBatch)
+		s.flushFrequency = cfg.FlushFrequency
+		return actor.WorkerContinue
+	default:
+	}
+
+	select {
+	case <-ctx.Done():
+		return actor.WorkerEnd
+	case item, ok := <-s.inbox.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		err := s.Append(ctx, item)
+		if err != nil {
+			level.Error(s.logger).Log("msg", "unable to append to serializer", "err", err)
+		}
+		return actor.WorkerContinue
+	case item, ok := <-s.metaInbox.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		err := s.AppendMetadata(ctx, item)
+		if err != nil {
+			level.Error(s.logger).Log("msg", "unable to append metadata to serializer", "err", err)
+		}
+		return actor.WorkerContinue
+	case <-s.flushTestTimer.C:
+		if time.Since(s.lastFlush) > s.flushFrequency {
+			err := s.store(ctx)
+			if err != nil {
+				level.Error(s.logger).Log("msg", "unable to store data", "err", err)
+			}
+		}
+		return actor.WorkerContinue
+	}
+}
+
+func (s *serializer) AppendMetadata(ctx actor.Context, data *types.TimeSeriesBinary) error {
+	s.meta = append(s.meta, data)
+	// If we would go over the max size then send, or if we have hit the flush duration then send.
+	if len(s.meta)+len(s.series) >= s.maxItemsBeforeFlush {
+		return s.store(ctx)
+	} else if time.Since(s.lastFlush) > s.flushFrequency {
+		return s.store(ctx)
+	}
+	return nil
+}
+
+func (s *serializer) Append(ctx actor.Context, data *types.TimeSeriesBinary) error {
```

Comment:
I would prefer to call this something more descriptive, like `storeInMem` - they also don't need to be exported AFAICT?

Reply:
Lemme refactor this out entirely, and the func can be moved to the do work loop.